presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java

```diff
@@ -0,0 +1,240 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.reservoirsample;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.DoubleStream;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class WeightedDoubleReservoirSample
+        implements Cloneable
+{
+    private static final int MAX_SAMPLES_LIMIT = 1_000_000;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(WeightedDoubleReservoirSample.class).instanceSize();
+
+    private int size;
+    private double[] samples;
+    private double[] weights;
+
+    public WeightedDoubleReservoirSample(int maxSamples)
+    {
+        checkArgument(maxSamples > 0, "Maximum number of reservoir samples must be positive");
+        checkArgument(maxSamples <= MAX_SAMPLES_LIMIT, "Maximum number of reservoir samples is capped");
+
+        this.size = 0;
+        this.samples = new double[maxSamples];
+        this.weights = new double[maxSamples];
+        checkArguments();
+    }
+
+    private WeightedDoubleReservoirSample(WeightedDoubleReservoirSample other)
+    {
+        checkArgument(other.size <= other.samples.length, "size must not be larger than samples");
+        this.size = other.size;
+        this.samples = new double[other.samples.length];
+        this.weights = new double[other.weights.length];
+        System.arraycopy(other.samples, 0, samples, 0, size);
+        System.arraycopy(other.weights, 0, weights, 0, size);
+        checkArguments();
+    }
+
+    private WeightedDoubleReservoirSample(int size, double[] samples, double[] weights)
+    {
+        this.size = size;
+        this.samples = samples;
+        this.weights = weights;
+    }
+
+    public long getMaxSamples()
+    {
+        return samples.length;
+    }
+
+    public void add(double element, double weight)
+    {
+        addWithAdjustedWeight(element, getAdjustedWeight(weight));
+    }
+
+    private void addWithAdjustedWeight(double element, double adjustedWeight)
+    {
+        if (size < samples.length) {
+            samples[size] = element;
+            weights[size] = adjustedWeight;
+            ++size;
+            bubbleUp();
+            checkArguments();
+            return;
+        }
+
+        if (adjustedWeight <= weights[0]) {
```

Comment:
it is confusing that `count` is not incremented here; is this why you names this variable `size`?

Reply:
This refers just to the count of samples stored in the array. If this line is reached, the number of samples doesn't change, only possibly _which_ samples are stored.